src/Lecture2.hs

```diff
@@ -292,4 +377,21 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding expr = formatExpr constants variableExprs
+    where
+      formatExpr :: [Int] -> [Expr] -> Expr
+      formatExpr [] vars = format' vars
+      formatExpr lits [] = Lit (sum lits)
+      formatExpr lits vars = case sum lits of 
+        0 -> format' vars
+        x -> Add (format' vars) (Lit x)
+      format' (x:[]) = x
+      format' (x:xs) = Add x (format' xs)
+      (constants, variableExprs) = parseAcc ([], []) expr
+      parseAcc :: ([Int], [Expr]) -> Expr -> ([Int], [Expr])
+      parseAcc (lits, vars) expr = case expr of
+        Lit x               -> (lits ++ [x], vars)
```

Comment:
You can store `Int` directly here instead of `[Int]` and append numbers immediately to avoid extra allocations

Reply:
I see.

Reply:
You don't need `Maybe Int`. You already have enough information to decide whether you need to add `Lit` to the final result or not: you need to check only the combination of literal sum values (is it 0 or not?) and whether there're any string liters. So, only four possible scenarios to handle 🙂